### PR TITLE
Add mechanism for regex filtering

### PR DESCRIPTION
**Pull request by pvandyken**
Filters can now take a subkey: `match` or `search`, which will change the matching method from the default direct match to a python regex method.

### CODE REVIEW

File: snakebids/core/_querying.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+
+@attrs.define
+class _TooFewKeysError(FilterSpecError):
+    """Exception raised when filter specified as dictionary without keys."""
+
+    @override
+    def get_config_error(self, component_name: str) -> ConfigError:
+        msg = (
+            f"Filter '{self.entity}' for component '{component_name}'  was specified a "
```
Comment (kaitj) [inline snakebids/core/_querying.py:282]: There is an extra space here between "'{component_name}'" and "was"

File: snakebids/core/input_generation.py [hunk 2]
```diff
@@ -466,195 +469,191 @@ def write_derivative_json(snakemake: Snakemake, **kwargs: dict[str, Any]) -> Non
...
+    component: InputConfig,
+    *,
+    input_name: str,
+    postfilters: PostFilter,
+) -> BidsComponent | None:
+    """Create component based on provided config.
 
+    Parameters
+    ----------
+    bids_layout : BIDSLayout
```
Comment (kaitj) [inline snakebids/core/input_generation.py:529]: Don't think this type hint needs to be included here

File: snakebids/core/input_generation.py [hunk 3]
```diff
@@ -466,195 +469,191 @@ def write_derivative_json(snakemake: Snakemake, **kwargs: dict[str, Any]) -> Non
...
+        Layout from pybids for accessing the BIDS dataset to grab paths
 
-@attrs.define(slots=False)
-class _UnifiedFilter:
-    """Manages component level and post filters."""
+    pybids_inputs : dict
```
Comment (kaitj) [inline snakebids/core/input_generation.py:532]: Was this suppose to refer to `component`?

File: snakebids/core/input_generation.py [hunk 4]
```diff
@@ -466,195 +469,191 @@ def write_derivative_json(snakemake: Snakemake, **kwargs: dict[str, Any]) -> Non
...
 
-    Currently only used to implement --[exclude-]participant-label, but in the future,
-    may implement other such CLI args. Unlike configuration-defined filters, these
-    filters apply after the dataset is indexed and queried. Thus, if a filter is set
-    to an empty list, a complete, albeit empty, component may be found. This is akin to
-    calling ``BidsComponent.filter`` after running ``generate_inputs``.
+    Yields
+    ------
+    BidsComponent
+        One BidsComponent is yielded for each modality described by ``pybids_inputs``.
```
Comment (kaitj) [inline snakebids/core/input_generation.py:545]: Again `component` or referring to the passed in `pybids_inputs`?

File: snakebids/core/input_generation.py [hunk 5]
```diff
@@ -466,195 +469,191 @@ def write_derivative_json(snakemake: Snakemake, **kwargs: dict[str, Any]) -> Non
...
-        """
-        wildcards: list[str] = []
-        if postfilter is not None:
-            wildcards.extend(postfilter.inclusions)
-            wildcards.extend(postfilter.exclusions)
-        return cls(
-            {"filters": filters, "wildcards": wildcards}, postfilter or _Postfilter()
+    if "custom_path" in component:
+        # a custom path was specified for this input, skip pybids:
+        # get input_wildcards by parsing path for {} entries (using a set
```
Comment (kaitj) [inline snakebids/core/input_generation.py:558]: Minor comment / q - is using a set still true here?
This whole comment confuses me a little bit with what is happening.

**Review by kaitj — COMMENTED** [review]
Largely looks good to me, some comments on some of the code comments and docstring.

File: snakebids/core/input_generation.py (hunk 5 above)
Comment (pvandyken) [reply snakebids/core/input_generation.py:558]: I think that comment's pretty ancient, I'll just delete it

**Review by kaitj — APPROVED** [review]

**Pull request merged**